---
title: "United States Scalloped Hammerhead Nurseries, Gene Diversity"
output: 
  html_document: 
    toc: yes
  html_notebook: 
    highlight: kate
    theme: flatly
    toc: yes
---

Assess genetic diversity of neutral and putatively adaptive loci by nursery.

# Environment
  
```{r, message=FALSE}

.libPaths("/usr/lib64/R/library")

# invalidate cache when the package version changes

knitr::opts_chunk$set(
  root.dir = "~/projects/nurseries",
	message = FALSE,
	warning = FALSE,
  cache.extra = packageVersion("tint"),
	tidy = FALSE,
	echo = FALSE)

options(htmltools.dir.version = FALSE)

# conflicts

library(conflicted)
conflict_prefer("count", "dplyr")
conflict_prefer("arrange", "dplyr")
conflict_prefer("mutate", "dplyr")
conflict_prefer("select", "dplyr")
conflict_prefer("filter", "dplyr")
conflict_prefer("rename", "dplyr")
conflict_prefer("summarise", "dplyr")
conflict_prefer("summarize", "dplyr")
conflict_prefer("s.label", "adegraphics")
conflict_prefer("s.value", "adegraphics")
conflict_prefer("scalebar", "raster")
conflict_prefer("rename", "dplyr")
conflict_prefer("extract", "raster")
conflict_prefer("degree", "igraph")

# packages

library(tidyverse)
library(adegenet)
library(here)
library(readr)
library(glue)
library(knitr)
library(poppr)
library(hierfstat)
library(coin)
library(zvau)
library(gdata)
library(readxl)
library(Hmisc)
library(cowplot)
library(vcfR)
library(haplotypes)
library(pegas)

# source scripts

source("~/bin/ggplot.R")
source("~/bin/genind.R")
source("~/bin/PCA.R")
source("~/bin/DAPC.R")
source("~/bin/vegan.R")
source("~/bin/ordiR2step.R")

# orders, colors, shapes

nursery_order <- c("BB", "TR", "CB", "FPH", "CCB")

nursery_order_full <- c("Bulls Bay", "Tolomator River", "Cape Canaveral", "Florida Panhandle", "Corpus Christi Bay")

nursery_map <- rlang::set_names(nursery_order_full, nursery_order)

nursery_colors <- c('#a50026', '#f46d43', '#fee090', '#74add1', '#313695')

```

# No Kin

```{r, message=F}

slew_no_sibs.gen <- read.genepop(file = here("results", "slew_no_sibs.gen"), ncode = 3L, quiet = FALSE)

# strata

slew_strata <- read_csv(here("results", "slew_no_sibs_strata.csv")) %>% 
  mutate(
    Nursery = dplyr::recode(nursery, !!!nursery_map),
    Nursery = factor(Nursery, levels = nursery_order_full)
  )

slew_strata <- indNames(slew_no_sibs.gen) %>%
  as_tibble_col("seq_id") %>% 
  left_join(., slew_strata) %>% 
  mutate(nursery = ordered(nursery, levels = nursery_order))

strata(slew_no_sibs.gen) <- slew_strata

setPop(slew_no_sibs.gen) <- ~nursery

slew_nurseries <- seppop(slew_no_sibs.gen)

```

Calculate diversity stats.

```{r, message=F}

loci_stats <- list()

for (g in names(slew_nurseries)) {

loc1 <- locus_table(slew_nurseries[[g]], index = "shannon") %>%
  as.data.frame() %>%
  rownames_to_column("locus") %>% 
  mutate(nursery = g)

loc2 <- locus_table(slew_nurseries[[g]], index = "simpson") %>%
  as.data.frame() %>%
  rownames_to_column("locus")

temp <- left_join(loc1, loc2)

loc3 <- locus_table(slew_nurseries[[g]], index = "invsimpson") %>%
  as.data.frame() %>%
  rownames_to_column("locus")

loci_stats[[g]] <- left_join(temp, loc3)

}

loci_stats_df <- tibble(do.call(rbind, loci_stats)) %>% 
  rename(simpson = "1-D",
         n_alleles = allele,
         shannon = H,
         stood_taylor = G,
         evenness = Evenness) %>%
  filter(locus != "mean") %>% 
  relocate(nursery, .after=locus) %>% 
  mutate(
    Nursery = dplyr::recode(nursery, !!!nursery_map),
    Nursery = factor(Nursery, levels = nursery_order_full)
  ) %>% 
  write_csv(., here("data", "diversity", "slew_nursery_no_sibs_neu_loci_stats.csv"))

```

## Nei's Gene Diversity

Compare patterns of Nei's gene diversity (i.e., expected heterozygosity) across loci by nursery.

```{r fig.height=5, fig.width=10, message=FALSE, warning=FALSE}

slew_nursery_no_sibs_neu_loci_stats <- read_csv(here("data", "diversity", "slew_nursery_no_sibs_neu_loci_stats.csv"))

slew_nursery_no_sibs_neu_nei <- slew_nursery_no_sibs_neu_loci_stats %>%
  select(nursery, Nursery, locus, Hexp) %>% 
  mutate(Nursery = ordered(Nursery, levels = nursery_order_full))

slew_nursery_no_sibs_neu_nei_mean <- slew_nursery_no_sibs_neu_nei %>%
  group_by(nursery) %>%
  summarise(mean(Hexp), sd(Hexp)) %>% 
  mutate(nursery = ordered(nursery, levels = nursery_order)) %>% 
  write_csv(., here("data", "diversity", "slew_nursery_no_sibs_neu_nei_mean.csv"))

# plot

ggplot(slew_nursery_no_sibs_neu_nei) +
  geom_violin(aes(x = Nursery, y = Hexp, fill = Nursery), show.legend = F) +
  scale_fill_manual(values = nursery_colors) +
  scale_y_continuous("Nei's Gene Diversity (Neutral Loci)", limits = c(0, 1), breaks = c(seq(0, 1, 0.1)), labels = c(seq(0, 1, 0.1))) +
  theme_standard

ggsave(here("data", "diversity", "slew_nursery_no_sibs_neu_nei.png"))

```

Test for significant differences among nurseries bootstrapping.

```{r fig.height=5, fig.width=10, message=FALSE, warning=FALSE}

boots <- 1e4
set.seed(1)

nei_boot_means <- slew_nursery_no_sibs_neu_nei %>%
  select(Nursery, locus, Hexp) %>%
  filter(!is.na(Hexp)) %>%
  group_by(Nursery) %>%
  group_modify(~ {
    x <- .x$Hexp
    n <- length(x)

    # build indices as n×boots, then coerce x[idx] into the same shape
    idx   <- matrix(sample.int(n, n * boots, replace = TRUE), nrow = n, ncol = boots)
    x_mat <- matrix(x[idx], nrow = n, ncol = boots)
    tibble(
      repl     = seq_len(boots),
      mean_nei = colMeans(x_mat)
    )
  }) %>%
  ungroup()

nei_obs_means <- slew_nursery_no_sibs_neu_nei %>%
  filter(!is.na(Hexp)) %>%
  group_by(Nursery) %>%
  summarise(obs_mean = mean(Hexp), .groups = "drop")

nei_nursery_cis <- nei_boot_means %>%
  group_by(Nursery) %>%
  summarise(
    boot_mean = mean(mean_nei),
    q025 = quantile(mean_nei, 0.025),
    q975 = quantile(mean_nei, 0.975),
    .groups = "drop"
  ) %>%
  left_join(nei_obs_means, by = "Nursery")

# plot 

ggplot(nei_nursery_cis, aes(x = Nursery, y = obs_mean)) +
  geom_point() +
  geom_errorbar(aes(ymin = q025, ymax = q975), width = 0.2) +
  ylab("Nei's Gene Diversity") +
  theme_standard

```

There is some variation among nurseries, but all of the confidence intervals overlap.

## Rarefied Allele Richness

Calculate allelic richness corrected for sample size using rarefaction.

```{r}

# by nursery

setPop(slew_no_sibs.gen) <- ~nursery

gen_est <- seppop(slew_no_sibs.gen)

temp <- gen_est[c(nursery_order)]

temp <- repool(temp)

setPop(temp) <- ~nursery

dat <- genind2hierfstat(temp)

df <- allelic.richness(dat, diploid = TRUE)

ar <- as.data.frame(df$Ar) %>%
  rownames_to_column("locus") %>% 
  write_csv(., here("data", "diversity", "slew_nursery_no_sibs_neu_ar_locus.csv"))

```

Compare patterns of rarefied allelic richness across loci by nursery.

```{r fig.height=5, fig.width=10, message=FALSE, warning=FALSE}

slew_nursery_no_sibs_neu_ar <- read_csv(here("data", "diversity", "slew_nursery_no_sibs_neu_ar_locus.csv")) %>%
  gather(key = nursery, value = AR, 2:(length(nursery_order)+1)) %>% 
  mutate(
    Nursery = dplyr::recode(nursery, !!!nursery_map),
    Nursery = factor(Nursery, levels = nursery_order_full)
  ) %>% 
  mutate(Nursery = ordered(Nursery, levels = nursery_order_full))

slew_nursery_no_sibs_neu_ar_mean <- slew_nursery_no_sibs_neu_ar %>%
  group_by(Nursery) %>%
  summarise(mean(AR), sd(AR)) %>% 
  rename(mean_AR = "mean(AR)", sd_AR = "sd(AR)") %>% 
  mutate(Nursery = ordered(Nursery, levels = nursery_order_full)) %>% 
  write_csv(., here("data", "diversity", "slew_nursery_no_sibs_neu_ar_mean.csv"))

# plot

ggplot(slew_nursery_no_sibs_neu_ar) +
  geom_violin(aes(x = Nursery, y = AR, fill = Nursery), show.legend = F) +
  scale_fill_manual(values = nursery_colors) +
  scale_y_continuous("Rarefied Allelic Richness (Neutral Loci)", limits = c(1, 6.5), breaks = c(seq(1, 6, 1)), labels = c(seq(1, 6, 1))) +
  guides(fill = guide_legend(nrow = 5, direction = "vertical", reverse = TRUE)) +
  theme_standard

ggsave(here("data", "diversity", "slew_nursery_no_sibs_neu_ar.png"))

```

Test for significant differences among nurseries bootstrapping.

```{r fig.height=5, fig.width=10, message=FALSE, warning=FALSE}

boots <- 1e4            # number of bootstrap replicates (tune as needed)
set.seed(1)   # reproducibility

ar_boot_means <- slew_nursery_no_sibs_neu_ar %>%
  select(Nursery, locus, AR) %>%
  filter(!is.na(AR)) %>%
  group_by(Nursery) %>%
  group_modify(~ {
    x <- .x$AR
    n <- length(x)

    # build indices as n×boots, then coerce x[idx] into the same shape
    idx   <- matrix(sample.int(n, n * boots, replace = TRUE), nrow = n, ncol = boots)
    x_mat <- matrix(x[idx], nrow = n, ncol = boots)
    tibble(
      repl     = seq_len(boots),
      mean_ar = colMeans(x_mat)
    )
  }) %>%
  ungroup()

ar_obs_means <- slew_nursery_no_sibs_neu_ar %>%
  filter(!is.na(AR)) %>%
  group_by(Nursery) %>%
  summarise(obs_mean = mean(AR), .groups = "drop")

ar_nursery_cis <- ar_boot_means %>%
  group_by(Nursery) %>%
  summarise(
    boot_mean = mean(mean_ar),
    q025 = quantile(mean_ar, 0.025),
    q975 = quantile(mean_ar, 0.975),
    .groups = "drop"
  ) %>%
  left_join(ar_obs_means, by = "Nursery")

# plot 

ggplot(ar_nursery_cis, aes(x = Nursery, y = obs_mean)) +
  geom_point() +
  geom_errorbar(aes(ymin = q025, ymax = q975), width = 0.2) +
  ylab("Rarefied Allelic Richness") +
  theme_standard

```

## Nucleotide Diversity (pi)

```{r}

# import fst outliers to remove

fst_outliers <- read_csv(here("data", "fst_outlier", "slew_no_sibs_outliers.csv"))

# miseq reference fasta

ref_fa <- seqinr::read.fasta(here("results", "reference.fasta"))

# create tibble with locus name and sequence

ref <- tibble::tibble(name = seqinr::getName(ref_fa), seq = toupper(unlist(seqinr::getSequence(ref_fa, as.string = TRUE))))

# import vcf that corresponds to genepop

vcf_tidy <- read.vcfR(here("data", "fst_outlier", "slew_nurseries_no_sibs.recode.vcf")) %>%
  vcfR2tidy()

# produce df with SNP positions

pos_tbl <- vcf_tidy$fix %>%
  filter(ALT %in% c("A", "T", "C", "G"))

gen_obj <- slew_no_sibs.gen

# Make a data frame with all of the haplotype sequences from the 'codes.out' file from rad_haplotyper

hap_index <- read_tsv(here("data", "haplotype", "codes.slew_nurseries.gen"), col_names = FALSE) %>%
  filter(X1 %in% locNames(gen_obj)) %>%
  split(.$X1) %>%
  purrr::map(function(x) {
    codes <- str_split(x$X2, ",") %>%
      unlist()

    code_tbl <- tibble(locus = x$X1, code = codes) %>%
      separate(code, c("hap", "code"), sep = ":") %>%
      left_join(ref, by = c("locus" = "name"))

    pos <- pos_tbl %>%
      filter(CHROM == code_tbl$locus[1]) %>%
      pull(POS)

    for (i in 1:nrow(code_tbl)) {

      alleles <- str_split(code_tbl$hap[i], "") %>%
        unlist()

      replace <- tibble(pos = pos, allele = alleles)

      for (j in 1:nrow(replace)) {
        str_sub(code_tbl$seq[i], replace$pos[j], 1) <- replace$allele[j]
      }

    }

    code_tbl

  }) %>%
  bind_rows()

# create tidy data set
gen_tidy <- gen_obj %>%
  genind2df(oneColPerAll = TRUE) %>%
  rownames_to_column(var = "ind") %>%
  gather(allele, code, -pop, -ind) %>%
  separate(allele, c("locus", "allele"), sep = "\\.") %>%
  #filter(pop %in% c("SanAntonioBay", "SabineLake",
  #                "BaratariaBay", "WestMississippiSound", "MobileBay",
  #                "ApalachicolaBay",
  #                "StJohnsRiver", "CharlestonHarbor", "WinyahBay")) %>%
  droplevels()

# Calculate haplotype related stats
hap_div_tbl <- locNames(gen_obj) %>%
  purrr::map(function(y) {

    # for each locus
    gen_tidy %>%
      filter(locus == y) %>%
      filter(!code == "NA") %>%
      left_join(hap_index, by = c("code", "locus")) %>%
      as.tibble() %>%
      arrange(factor(pop, levels = nursery_order)) %>% 
      # for each population per locus
      split(.$pop) %>%
      purrr::map(function(x) {
        y <- do.call(rbind, strsplit(x$seq, ""))
        rownames(y) <- paste(x$ind, x$allele, sep = ".")

        # create DNAbin object for locus
        dna_bin <- haplotypes::as.DNAbin(y)                            

        # calculate Tajima's D and p-values
        #taj_test <- tajima.test(dna_bin)

        # create table with results
        tibble(locus = x$locus[1],
               pop = x$pop[1],
               hap_div = hap.div(dna_bin),                     # calculate haplotype diversity
               nuc_div = nuc.div(dna_bin)                     # calculate nucleotide diversity (pi)
               #seg_sites = length(seg.sites(dna_bin)),         # count number of segregating sites
               #thetaS = theta.s(x = seg_sites, n = nLoc(slew_no_sibs.gen)), # calculate theta S (watterson 1975)
               #tajima_d = taj_test$D,                          # extract Tajima's D
               #tajima_pval_norm = taj_test$Pval.normal,        # extract p-value Tajima's D (normal distribution)
               #tajima_pval_beta = taj_test$Pval.beta          # extract p-value Tajima's D (beta distribution)
               )          
      }) %>%
      bind_rows()
  }) %>%
    bind_rows() %>% 
  write_csv(., here("data", "diversity", "slew_nurseries_hap_nuc_diversity.csv"))

```

Calculate mean and bootstrapped confidence interval values, then plot.

```{r fig.height=5, fig.width=10, message=FALSE, warning=FALSE}

slew_nursery_no_sibs_neu_pi <- read_csv(here("data", "diversity", "slew_nurseries_hap_nuc_diversity.csv")) %>%
 # gather(key = nursery, value = AR, 2:(length(nursery_order)+1)) %>% 
  rename(nursery = pop) %>% 
  mutate(
    Nursery = dplyr::recode(nursery, !!!nursery_map),
    Nursery = factor(Nursery, levels = nursery_order_full)
  ) %>% 
  mutate(Nursery = ordered(Nursery, levels = nursery_order_full))

slew_nursery_no_sibs_neu_pi_mean <- slew_nursery_no_sibs_neu_pi %>%
  group_by(Nursery) %>%
  summarise(mean(nuc_div), sd(nuc_div)) %>% 
  rename(mean_pi = "mean(nuc_div)", sd_pi = "sd(nuc_div)") %>% 
  mutate(Nursery = ordered(Nursery, levels = nursery_order_full)) %>% 
  write_csv(., here("data", "diversity", "slew_nursery_no_sibs_neu_pi_mean.csv"))

# plot

ggplot(slew_nursery_no_sibs_neu_pi) +
  geom_violin(aes(x = Nursery, y = nuc_div, fill = Nursery), show.legend = F) +
  scale_fill_manual(values = nursery_colors) +
  scale_y_continuous(
    "Nucleotide Diversity (Neutral Loci)", 
    #limits = c(1, 6.5), 
    #breaks = c(seq(1, 6, 1)), 
    #labels = c(seq(1, 6, 1)
    ) +
  guides(fill = guide_legend(nrow = 5, direction = "vertical", reverse = TRUE)) +
  theme_standard

ggsave(here("data", "diversity", "slew_nursery_no_sibs_neu_pi.png"))

```

Test for significant differences among nurseries bootstrapping.

```{r fig.height=5, fig.width=10, message=FALSE, warning=FALSE}

boots <- 1e4            # number of bootstrap replicates (tune as needed)
set.seed(1)   # reproducibility

pi_boot_means <- slew_nursery_no_sibs_neu_pi %>%
  select(Nursery, locus, nuc_div) %>%
  filter(!is.na(nuc_div)) %>%
  group_by(Nursery) %>%
  group_modify(~ {
    x <- .x$nuc_div
    n <- length(x)

    # build indices as n×boots, then coerce x[idx] into the same shape
    idx   <- matrix(sample.int(n, n * boots, replace = TRUE), nrow = n, ncol = boots)
    x_mat <- matrix(x[idx], nrow = n, ncol = boots)
    tibble(
      repl     = seq_len(boots),
      mean_pi = colMeans(x_mat)
    )
  }) %>%
  ungroup()

pi_obs_means <- slew_nursery_no_sibs_neu_pi %>%
  filter(!is.na(nuc_div)) %>%
  group_by(Nursery) %>%
  summarise(obs_mean = mean(nuc_div), .groups = "drop")

pi_nursery_cis <- pi_boot_means %>%
  group_by(Nursery) %>%
  summarise(
    boot_mean = mean(mean_pi),
    q025 = quantile(mean_pi, 0.025),
    q975 = quantile(mean_pi, 0.975),
    .groups = "drop"
  ) %>%
  left_join(pi_obs_means, by = "Nursery")

# plot 

ggplot(pi_nursery_cis, aes(x = Nursery, y = obs_mean)) +
  geom_point() +
  geom_errorbar(aes(ymin = q025, ymax = q975), width = 0.2) +
  ylab("Nucleotide Diversity") +
  theme_standard

```

## FIS

```{r fig.height=5, fig.width=10, message=FALSE, warning=FALSE}

hf <- genind2hierfstat(slew_no_sibs.gen)
bs <- hierfstat::basic.stats(hf)

fis_long <- as_tibble(bs$Fis, rownames = "locus") %>%
  pivot_longer(-locus, names_to = "nursery", values_to = "FIS") %>%
  mutate(
    Nursery = dplyr::recode(nursery, !!!nursery_map),
    Nursery = factor(Nursery, levels = nursery_order_full)
  ) %>% 
  mutate(Nursery = ordered(Nursery, levels = nursery_order_full))

fis_summ <- fis_long %>%
  group_by(nursery) %>%
  summarise(
    FIS_mean   = mean(FIS, na.rm = TRUE),
    FIS_median = median(FIS, na.rm = TRUE),
    n_loci     = sum(is.finite(FIS)),
    .groups = "drop"
  ) %>% 
  mutate(
    Nursery = dplyr::recode(nursery, !!!nursery_map),
    Nursery = factor(Nursery, levels = nursery_order_full)
  ) %>% 
  mutate(Nursery = ordered(Nursery, levels = nursery_order_full))

boots <- 1e4
set.seed(1)

fis_boot_means <- fis_long %>%
  filter(is.finite(FIS)) %>%
  group_by(Nursery) %>%
  group_modify(~ {
    x <- .x$FIS
    n <- length(x)
    if (n < 2L) {
      return(tibble(repl = integer(), mean_fis = numeric()))
    }
    idx   <- matrix(sample.int(n, n * boots, replace = TRUE), nrow = n, ncol = boots)
    x_mat <- matrix(x[idx], nrow = n, ncol = boots)
    tibble(repl = seq_len(boots), mean_fis = colMeans(x_mat))
  }) %>%
  ungroup()

fis_obs_means <- fis_long %>%
  filter(is.finite(FIS)) %>%
  group_by(Nursery) %>%
  summarise(obs_mean = mean(FIS), .groups = "drop")

fis_nursery_cis <- fis_boot_means %>%
  group_by(Nursery) %>%
  summarise(
    boot_mean = mean(mean_fis),
    q025 = quantile(mean_fis, 0.025),
    q975 = quantile(mean_fis, 0.975),
    .groups = "drop"
  ) %>%
  right_join(fis_obs_means, by = "Nursery") %>%     # keep nurseries even if no CI (n < 2)
  left_join(fis_summ, by = "Nursery") %>% 
  mutate(Nursery = ordered(Nursery, levels = nursery_order_full))

# plot 

ggplot(fis_nursery_cis, aes(x = Nursery, y = obs_mean)) +
  geom_point() +
  geom_errorbar(aes(ymin = q025, ymax = q975), width = 0.2) +
  ylab("FIS") +
  theme_standard

# 4) Which nurseries have non-overlapping CIs?
# Define overlap: intervals overlap unless one upper < the other's lower.
ci_bounds <- fis_nursery_cis %>%
  transmute(Nursery, lo = q025, hi = q975) %>%
  filter(is.finite(lo), is.finite(hi))

idx <- ci_bounds %>%
  transmute(Nursery, i = row_number())

pairs_df <- idx %>%
  rename(nursery1 = Nursery, i1 = i) %>%
  tidyr::crossing(idx %>% rename(nursery2 = Nursery, i2 = i)) %>%
  filter(i1 < i2) %>%
  select(nursery1, nursery2) %>%
  left_join(ci_bounds, by = c("nursery1" = "Nursery")) %>%
  rename(lo1 = lo, hi1 = hi) %>%
  left_join(ci_bounds, by = c("nursery2" = "Nursery")) %>%
  rename(lo2 = lo, hi2 = hi) %>%
  mutate(
    overlap     = !(hi1 < lo2 | hi2 < lo1),  # touching counts as overlap
    non_overlap = !overlap
  )

# Optional: a compact list of clearly separated pairs
nonoverlap_pairs <- pairs_df %>%
  filter(non_overlap) %>%
  arrange(nursery1, nursery2)

```
